packages/datetime/src/dateInput.tsx

```diff
@@ -314,7 +316,11 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
                 Utils.safeInvoke(this.props.onChange, this.fromMomentToDate(value));
             }
         } else {
-            this.setState({ isInputFocused: false });
+            if (!valueString.length) {
+                this.setState({ isInputFocused: false, value: moment(null), valueString: null });
```

Comment:
I tried to follow [your style](https://github.com/palantir/blueprint/blob/master/packages/datetime/src/dateInput.tsx#L304)